File: src/brs/web/api/http/handler/EscrowSign.java
package brs.web.api.http.handler;

import brs.*;
import brs.services.EscrowService;
import brs.services.ParameterService;
import brs.util.Convert;
import brs.web.api.http.common.APITransactionManager;
import brs.web.api.http.common.LegacyDocTag;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import javax.servlet.http.HttpServletRequest;

import static brs.web.api.http.common.Parameters.DECISION_PARAMETER;
import static brs.web.api.http.common.Parameters.ESCROW_PARAMETER;
import static brs.web.api.http.common.ResultFields.ERROR_CODE_RESPONSE;
import static brs.web.api.http.common.ResultFields.ERROR_DESCRIPTION_RESPONSE;

public final class EscrowSign extends CreateTransaction {

  private final ParameterService parameterService;
  private final EscrowService escrowService;
  private final Blockchain blockchain;

  public EscrowSign(ParameterService parameterService, Blockchain blockchain, EscrowService escrowService, APITransactionManager apiTransactionManager) {
    super(new LegacyDocTag[] {LegacyDocTag.TRANSACTIONS, LegacyDocTag.CREATE_TRANSACTION}, apiTransactionManager, ESCROW_PARAMETER, DECISION_PARAMETER);
    this.parameterService = parameterService;
    this.blockchain = blockchain;
    this.escrowService = escrowService;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) throws SignumException {
    long escrowId;
    try {
      escrowId = Convert.parseUnsignedLong(Convert.emptyToNull(req.getParameter(ESCROW_PARAMETER)));
    }
    catch(Exception e) {
      JsonObject response = new JsonObject();
      response.addProperty(ERROR_CODE_RESPONSE, 3);
      response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Invalid or not specified escrow");
      return response;
    }

    Escrow escrow = escrowService.getEscrowTransaction(escrowId);
    if(escrow == null) {
      JsonObject response = new JsonObject();
      response.addProperty(ERROR_CODE_RESPONSE, 5);
      response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Escrow transaction not found");
      return response;
    }

    Escrow.DecisionType decision = Escrow.stringToDecision(req.getParameter(DECISION_PARAMETER));
    if(decision == null) {
      JsonObject response = new JsonObject();
      response.addProperty(ERROR_CODE_RESPONSE, 5);
      response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Invalid or not specified action");
      return response;
    }

    Account sender = parameterService.getSenderAccount(req);
    if(! isValidUser(escrow, sender)) {
      JsonObject response = new JsonObject();
      response.addProperty(ERROR_CODE_RESPONSE, 5);
      response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Invalid or not specified action");
      return response;
    }

    if(escrow.getSenderId().equals(sender.getId()) && decision != Escrow.DecisionType.RELEASE) {
      JsonObject response = new JsonObject();
      response.addProperty(ERROR_CODE_RESPONSE, 4);
      response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Sender can only release");
      return response;
    }

    if(escrow.getRecipientId().equals(sender.getId()) && decision != Escrow.DecisionType.REFUND) {
      JsonObject response = new JsonObject();
      response.addProperty(ERROR_CODE_RESPONSE, 4);
      response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Recipient can only refund");
      return response;
    }

    Attachment.AdvancedPaymentEscrowSign attachment = new Attachment.AdvancedPaymentEscrowSign(escrow.getId(), decision, blockchain.getHeight());

    return createTransaction(req, sender, null, 0, attachment);
  }

  private boolean isValidUser(Escrow escrow, Account sender) {
    return
        escrow.getSenderId().equals(sender.getId()) ||
        escrow.getRecipientId().equals(sender.getId()) ||
        escrowService.isIdSigner(sender.getId(), escrow);
  }
}

File: src/brs/http/EscrowSign.kt
package brs.http

import brs.*
import brs.services.EscrowService
import brs.services.ParameterService
import brs.util.Convert
import com.google.gson.JsonElement
import com.google.gson.JsonObject

import javax.servlet.http.HttpServletRequest

import brs.http.common.Parameters.DECISION_PARAMETER
import brs.http.common.Parameters.ESCROW_PARAMETER
import brs.http.common.ResultFields.ERROR_CODE_RESPONSE
import brs.http.common.ResultFields.ERROR_DESCRIPTION_RESPONSE

internal class EscrowSign internal constructor(private val parameterService: ParameterService, private val blockchain: Blockchain, private val escrowService: EscrowService, apiTransactionManager: APITransactionManager) : CreateTransaction(arrayOf(APITag.TRANSACTIONS, APITag.CREATE_TRANSACTION), apiTransactionManager, ESCROW_PARAMETER, DECISION_PARAMETER) {

    @Throws(BurstException::class)
    internal override fun processRequest(req: HttpServletRequest): JsonElement {
        val escrowId: Long
        try {
            escrowId = Convert.parseUnsignedLong(Convert.emptyToNull(req.getParameter(ESCROW_PARAMETER)))
        } catch (e: Exception) {
            val response = JsonObject()
            response.addProperty(ERROR_CODE_RESPONSE, 3)
            response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Invalid or not specified escrow")
            return response
        }

        val escrow = escrowService.getEscrowTransaction(escrowId)
        if (escrow == null) {
            val response = JsonObject()
            response.addProperty(ERROR_CODE_RESPONSE, 5)
            response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Escrow transaction not found")
            return response
        }

        val decision = Escrow.stringToDecision(req.getParameter(DECISION_PARAMETER))
        if (decision == null) {
            val response = JsonObject()
            response.addProperty(ERROR_CODE_RESPONSE, 5)
            response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Invalid or not specified action")
            return response
        }

        val sender = parameterService.getSenderAccount(req)
        if (!isValidUser(escrow, sender)) {
            val response = JsonObject()
            response.addProperty(ERROR_CODE_RESPONSE, 5)
            response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Invalid or not specified action")
            return response
        }

        if (escrow.getSenderId() == sender.getId() && decision != Escrow.DecisionType.RELEASE) {
            val response = JsonObject()
            response.addProperty(ERROR_CODE_RESPONSE, 4)
            response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Sender can only release")
            return response
        }

        if (escrow.getRecipientId() == sender.getId() && decision != Escrow.DecisionType.REFUND) {
            val response = JsonObject()
            response.addProperty(ERROR_CODE_RESPONSE, 4)
            response.addProperty(ERROR_DESCRIPTION_RESPONSE, "Recipient can only refund")
            return response
        }

        val attachment = Attachment.AdvancedPaymentEscrowSign(escrow.getId(), decision, blockchain.height)

        return createTransaction(req, sender, null, 0, attachment)
    }

    private fun isValidUser(escrow: Escrow, sender: Account): Boolean {
        return escrow.getSenderId() == sender.getId() ||
                escrow.getRecipientId() == sender.getId() ||
                escrowService.isIdSigner(sender.getId(), escrow)
    }
}
